internal/own/codeowners/find_owners.go

```
@@ -133,69 +165,114 @@ type globPattern []patternPart
 // The match is successful if after iterating through the whole file path,
 // full pattern matches, that is, there is a bit at the end of the glob.
 func (glob globPattern) match(filePath string) bool {
-	currentState := big.NewInt(0)
-	glob.markEmptyMatches(currentState)
-	filePathParts := strings.Split(strings.Trim(filePath, separator), separator)
-	nextState := big.NewInt(0)
-	for _, part := range filePathParts {
-		nextState.SetInt64(0)
-		glob.consume(part, currentState, nextState)
-		currentState, nextState = nextState, currentState
+	// Fast pass for literal globs, we can just string compare those.
+	if glob.isLiteral {
+		return glob.pattern == filePath
 	}
-	return glob.matchesWhole(currentState)
-}
+	// If starts with ** (ie no root match), do a fast pass on the last rule first,
+	// this optimizes file ending and file name matches.
+	if _, ok := glob.parts[glob.size-1].(anySubPath); ok {
+		l, ok := lastPart(filePath, '/')
+		if ok {
+			if !glob.parts[glob.size-1].Match(l) {
+				return false
+			}
+		}
+	}
+	// Dirty cheap version of strings.Trim(filePath, separator)
+	if filePath[0] == '/' {
+		filePath = filePath[1:]
+	}
+	if filePath[len(filePath)-1] == '/' {
+		filePath = filePath[:len(filePath)-1]
+	}
+	var (
+		currentState = glob.initialState
+		nextState    = int64(0)
+		part         string
+		hasNext      bool
+	)
+	for {
+		part, hasNext, filePath = nextPart(filePath, '/')
```

see docstring for func, but this implements a non-greedy strings.Split (with limitations, probably) that reduces allocations and memory usage for patterns that fail early.